rdfproxy/utils/utils.py

```diff
@@ -1,90 +1,76 @@
 """SPARQL/FastAPI utils."""
 
-from collections.abc import Iterator
-from typing import Any
+from collections.abc import Callable, Iterable
+from typing import Any, get_args, get_origin
 
-from SPARQLWrapper import QueryResult
 from pydantic import BaseModel
 from pydantic.fields import FieldInfo
-from rdfproxy.utils._types import SPARQLBinding, _TModelInstance
+from rdfproxy.utils._exceptions import (
+    MissingModelConfigException,
+    UnboundGroupingKeyException,
+)
+from rdfproxy.utils._types import SPARQLBinding
 
 
-def get_bindings_from_query_result(query_result: QueryResult) -> Iterator[dict]:
-    """Extract just the bindings from a SPARQLWrapper.QueryResult."""
-    if (result_format := query_result.requestedFormat) != "json":
-        raise Exception(
-            "Only QueryResult objects with JSON format are currently supported. "
-            f"Received object with requestedFormat '{result_format}'."
-        )
+def _type_p(obj: type | None, _type: type) -> bool:
+    """Check if an obj is type _type or a GenericAlias with origin _type."""
+    return (obj is _type) or (get_origin(obj) is _type)
 
-    query_json: dict = query_result.convert()
-    bindings = map(
-        lambda binding: {k: v["value"] for k, v in binding.items()},
-        query_json["results"]["bindings"],
-    )
-
-    return bindings
-
-
-def instantiate_model_from_kwargs(
-    model: type[_TModelInstance], **kwargs
-) -> _TModelInstance:
-    """Instantiate a (potentially nested) model from (flat) kwargs.
 
-    More a more generic version of this function see upto.init_model_from_kwargs
-    https://github.com/lu-pl/upto?tab=readme-ov-file#init_model_from_kwargs.
+def _list_p(obj: type | None) -> bool:
+    """Check if obj is a list type."""
+    return _type_p(obj, list)
```

Comment:
Please drop `_type_p` and put the code directly in `_list_p` - if its only used in this one method, thats too much abstraction

Reply:
Also, it would be great if the function name would be a bit more meaningful (i.e. add an `is`)

Reply:
True, the naming follows Lisp/Prolog convention for "p-redicates" and actually isn't a good choice for Python. I would change the names to `is_list_type` and `is_basemodel_type`, if that is alright.

I agree that `_type_p` is a case of somewhat unnecessary/premature abstraction, yet I would like to keep `_type_p` (renamed to `is_of_type`) separate because I think it is likely that there will be future type checking paths in `ModelBindingsMapper._generate_binding_pairs` that can reuse that function. 

I don't have a strong opinion about this though, so if you would like the `_type_p` logic in `_list_p`, I will implement it right away.

Reply:
Predicates are renamed to `_is_type`, `_is_list_type` and `_is_list_basemodel_type`.